# cuRRBS

Created by Daniel E. Martin-Herranz, Antonio J.M. Ribeiro and Thomas M. Stubbs.

Copyright (C) 2016,2017 D.E. Martin-Herranz, A.J.M. Ribeiro, T.M. Stubbs.

## What is cuRRBS?

**cuRRBS** (customised Reduced Representation Bisulfite Sequencing) is an easy-to-use computational method that predicts what is the optimal combination of *restriction enzymes* and *size range* that should be used in order to enrich for any given set of sites of interest in any genome. In other words, by modifying two steps of the original protocol, cuRRBS generalises [RRBS](http://www.nature.com/nprot/journal/v6/n4/full/nprot.2010.190.html). This allows to create 'personalised' reduced representations (i.e. subsets) of the genome, that make DNA methylation next-generation sequencing experiments more cost-effective.

## Getting started

Before being able to run cuRRBS in your computer, you will need to download and install the following software:

1. [Python 2.7](https://www.python.org/downloads/). The programming language in which cuRRBS is written.
2. [NumPy](https://scipy.org/install.html#individual-packages). A Python module which contains many useful functions used by cuRRBS.
3. [Terminal](https://en.wikipedia.org/wiki/Comparison_of_terminal_emulators). You will need it to run cuRRBS commands. 

Afterwards, you can install cuRRBS in your computer by [direct download](https://github.com/demh/cuRRBS/archive/master.zip) or cloning it from this GitHub repository:

```
git clone https://github.com/demh/cuRRBS.git
```

## Preparing cuRRBS input files

cuRRBS requires the following files as input:

**1. Isoschizomer annotation file.** 

Restriction enzymes that cut the genome in the same way (i.e. generate the same fragment length distribution) are grouped in isoschizomer families. Furthermore, this file also contains information regarding the methylation-sensitivity (in a concrete genomic context) of the commercially-available restriction enzymes. These files can be found in [utils/isoschizomers_*_annotation.csv](https://github.com/demh/cuRRBS/tree/master/utils) but the user can also create his own file. By default, cuRRBS uses [isoschizomers_CpG_annotation.csv](https://github.com/demh/cuRRBS/blob/master/utils/isoschizomers_CpG_annotation.csv), which should be used when considering only methylation-sensitivity in CpG contexts (i.e. for most vertebrate genomes). 

**2. Pre-computed files.** 

They contain information for the *in silico* digestions of a genome with certain restriction enzymes. There are two ways to obtain these files:

* [Download them](http://www.ebi.ac.uk/~dem44/cuRRBS_pre_computed_files/) (recommended option for human, mouse and *Arabidopsis thaliana* genomes). Only the enzymes which belong to an isoschizomer family with CpG methylation-insensitive members (see [utils/isoschizomers_CpG_annotation.csv](https://github.com/demh/cuRRBS/blob/master/utils/isoschizomers_CpG_annotation.csv)) are available. They need to be extracted using the following command in the terminal:

   ```
   tar xvjf pre_computed_files_for_genome_of_interest.tar.bz2 /path/to/precomputed/files/folder/
   ```

* Generated by the user (recommended only when using a novel genome). The script [src/pre_compute_digestions.py](https://github.com/demh/cuRRBS/blob/master/src/pre_compute_digestions.py) can be used for these purposes:

   ```   
   python pre_compute_digestions.py enzymes_to_pre_compute.txt genome.fa working_directory
   ``` 

**3. Enzymes to check.** 

This file contains the isoschizomer families that should be considered by cuRRBS (i.e. those that contain at least one methylation-insensitive member). When working with CpG methylation-insensitive enzymes, the file [utils/enzymes_to_check_CpG.txt](https://github.com/demh/cuRRBS/blob/master/utils/enzymes_to_check_CpG.txt) should be used.

**4. Sites annotation file.** 

This file stores the information regarding the sites of interest (i.e. genomic coordinates) that you want to enrich for with the new customised protocol. This file needs to be created from scratch, since it will be different depending on the sites that need to be targeted. The file is provided with a CSV (comma-separated values) format, with each row containing the information for a site of interest and the following columns:  

* *Site_ID*: unique ID for the current site of interest. It can not contain the '_' or the ',' characters.

* *Chr*: chromosome where the current site of interest is located. The same chromosome names should be used as in the case of the pre-computed files.

* *Coordinate*: 1-based genomic coordinate where the current site of interest is located. Please make sure that the coordinates are based on the same genome assembly as the one used to generate the pre-computed files.

* *Weight*: in case there is a preference in recovering certain sites of interest, their weights should be higher. The weights are always positive and are used afterwards to calculate the *Score* (see **Interpreting cuRRBS output** section).

The first line of the sites annotation file must be a header containing the column names. Some examples of these type of files for different biological systems can be found in the [examples/](https://github.com/demh/cuRRBS/tree/master/examples) folder.


## 3. Running cuRRBS

There are two parameters that you definitely need to consider before running cuRRBS:

* *C_Score constant* (compulsory). It provides a threshold for the minimum *Score* that needs to be obtained in order to include a certain enzyme combination in the output. This is important since there is a trade-off between the number of sites of interest that will be sequenced and the cost associated with the new protocol (i.e. higher *C_Score* values will be associated by higher sequencing costs). We recommend running the software initially with a value of 0.25, which will force the reported customised protocols to be able to capture at least 25 % of the maximum *Score* (i.e. in those cases where all the sites weights are the same, this is equivalent to 25 % of the sites of interest).   

* *Experimental error* (default: 20 bp). You need to specify what is your estimated experimental error during the size selection step. In general, this error should be higher for protocols that use [AMPure XP beads](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-r92) as compared to [gel-slicing](http://www.nature.com/nprot/journal/v6/n4/full/nprot.2010.190.html). If you run cuRRBS with a higher experimental error, less size ranges will be checked and the software will be faster. This parameter will also influence the *robustness* of the protocol (higher experimental errors generally imply lower *robustness*). 

You can find more information regarding cuRRBS parameters in the help page of the software:

```
python cuRRBS.py -h
``` 

A tipical cuRRBS command (using the default parameters) would look like:

```
python /path/to/cuRRBS/cuRRBS.py -o /path/to/output/folder/ -p /path/to/hg38/pre-computed/files/folder/ -e /path/to/cuRRBS/utils/enzymes_to_check_CpG.txt -a /path/to/cuRRBS/examples/epigenetic_clock_human_hg38_sites_annotation.csv -r 75 -s 120 -c 0.25 -g 3088.286401
```


## 4. Interpreting cuRRBS output  

CSV file, can be opened with Excel. Following columns:
